src/window/empire.c

```diff
@@ -407,13 +409,16 @@ static void draw_foreground(void)
     draw_panel_buttons(city);
 }
 
+static int is_outside_map(int x, int y)
+{
+    return (x < data.x_min + 16 || x >= data.x_max - 16 ||
+            y < data.y_min + 16 || y >= data.y_max - 120);
+}
+
 static void determine_selected_object(const mouse *m)
 {
-    if (!m->left.went_down) {
-        return;
-    }
-    if (m->x < data.x_min + 16 || m->x >= data.x_max - 16 ||
-        m->y < data.y_min + 16 || m->y >= data.y_max - 120) {
+    if (!((!m->is_touch && m->left.went_down ) || (m->is_touch && m->left.went_up && !data.finished_scroll)) || is_outside_map(m->x, m->y)) {
```

Comment:
Just split this out to multiple if-statements or make it a small function. Multiple parentheses and not-operators make me nervous ;)

Reply:
This was an unneeded mess.